### ISSUE TITLE
CI: Cilium not getting ready after restart in `K8sFQDNTest Restart Cilium validate that FQDN is still working`

### ISSUE THREAD

**Issue by nebril** [cilium/cilium#8305]
https://jenkins.cilium.io/job/Ginkgo-CI-Tests-Pipeline/7347/testReport/Suite-k8s-1/14/K8sFQDNTest_Restart_Cilium_validate_that_FQDN_is_still_working/

**Comment by ianvernon** [comment]
Duplicate of #8234

**Comment by raybejjani** [comment]
I was looking at another instance of this failure (https://jenkins.cilium.io/job/Ginkgo-CI-Tests-Pipeline/7361/testReport/junit/Suite-k8s-1/10/K8sFQDNTest_Restart_Cilium_validate_that_FQDN_is_still_working/) and I noticed something a little odd.

[afcc81e4_K8sFQDNTest_Restart_Cilium_validate_that_FQDN_is_still_working.zip](https://github.com/cilium/cilium/files/3295949/afcc81e4_K8sFQDNTest_Restart_Cilium_validate_that_FQDN_is_still_working.zip)

The status on failure is:
```
	Cilium Pod: cilium-d6v82 	Endpoint: 209 	Identity: 10422	 State: ready
	Cilium Pod: cilium-d6v82 	Endpoint: 1089 	Identity: 101	 State: ready
	Cilium Pod: cilium-d6v82 	Endpoint: 1905 	Identity: 61121	 State: ready
	Cilium Pod: cilium-d6v82 	Endpoint: 2024 	Identity: 9854	 State: not-ready
	Cilium Pod: cilium-d6v82 	Endpoint: 3661 	Identity: 61121	 State: ready
	Cilium Pod: cilium-d6v82 	Endpoint: 3820 	Identity: 4	 State: ready
	Cilium Pod: cilium-qwhzr 	Endpoint: 177 	Identity: 39899	 State: ready
	Cilium Pod: cilium-qwhzr 	Endpoint: 1554 	Identity: 101	 State: ready
	Cilium Pod: cilium-qwhzr 	Endpoint: 2116 	Identity: 101	 State: ready
	Cilium Pod: cilium-qwhzr 	Endpoint: 2252 	Identity: 100	 State: ready
	Cilium Pod: cilium-qwhzr 	Endpoint: 2360 	Identity: 102	 State: ready
	Cilium Pod: cilium-qwhzr 	Endpoint: 3241 	Identity: 105	 State: ready
	Cilium Pod: cilium-qwhzr 	Endpoint: 3900 	Identity: 4	 State: ready
```

and if you look at the logs for cilium-d6v82 this line shows up:
`2019-06-15T17:41:55.654519913Z level=debug msg="Error regenerating endpoint: Skipping build due to invalid state: ready" code=Failure containerID=cc6b44681f datapathPolicyRevision=4 desiredPolicyRevision=4 endpointID=2024 endpointState=ready identity=9854 ipv4=10.10.0.89 ipv6="f00d::a0a:0:0:5a96" k8sPodName=default/app2-5c44ff87c-sz8zt policyRevision=4 subsys=endpoint type=200`

The code where this error is generated is https://github.com/cilium/cilium/blob/master/pkg/endpoint/policy.go#L273-L283 and it explicitly expects Waiting-For-Identity.

Is it possible that we have 2 enqueued builds and the later one no-ops due to this error. Then cilium status lists it as not-ready because there was an error in the endpoint info, but it is actually working fine?

**Comment by jrajahalme** [comment]
@raybejjani The code generating the log is checking for the `WaitingForIdentity` state in order to skip the state change in that case. In all other cases it tires to change the state to `Regenerating`, which is not allowed when the endpoint is already in the `Ready` State. I think your suspicion that this may involve two parallel builds of the same endpoint has merit.

**Comment by jrajahalme** [comment]
Also in https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/13855/execution/node/152/log/

**Comment by ianvernon** [comment]
@jrajahalme the issue that this was related to was actuall #8234 , not the following, which was observed in your PR:

```
 "app2-bf8789b4f-rnldl" cannot curl to "http://world1.cilium.test"
[2019-07-24T21:28:36.071Z]   Expected command: kubectl exec -n default app2-bf8789b4f-rnldl -- curl --path-as-is -s -D /dev/stderr --fail --connect-timeout 5 --max-time 8 http://world1.cilium.test -w "time-> DNS: '%{time_namelookup}(%{remote_ip})', Connect: '%{time_connect}',Transfer '%{time_starttransfer}', total '%{time_total}'" 
[2019-07-24T21:28:36.071Z]   To succeed, but it failed:
[2019-07-24T21:28:36.071Z]   Exitcode: 28 
[2019-07-24T21:28:36.071Z]   Stdout:
[2019-07-24T21:28:36.071Z]    	 time-> DNS: '0.029362()', Connect: '0.000000',Transfer '0.000000', total '5.003163'
[2019-07-24T21:28:36.071Z]   Stderr:
[2019-07-24T21:28:36.071Z]    	 command terminated with exit code 28
[2019-07-24T21:28:36.071Z]   	 
[2019-07-24T21:28:36.071Z]   
```

in #8234 , endpoints went into `not-ready` state after restoring. Did the same occur in the build to which you are referring?

**Comment by ianvernon** [comment]
Closing this; the issue referred to in @raybejjani 's comment has been fixed by the PR addressing #8234 . All other build links have been GC'd by Jenkins, so nothing actionable is possible here at this time.
